networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRangeMessageHandler.java

```diff
@@ -120,35 +120,43 @@ public void onIncomingMessage(
   }
 
   private SafeFuture<RequestState> sendNextBlock(final RequestState requestState) {
-    SafeFuture<Optional<SignedBeaconBlock>> blockFuture = requestState.loadNextBlock();
+    SafeFuture<Boolean> blockFuture = processNextBlock(requestState);
     // Avoid risk of StackOverflowException by iterating when the block future is already complete
     // Using thenCompose on the completed future would execute immediately and recurse back into
     // this method to send the next block.  When not already complete, thenCompose is executed
     // on a separate thread so doesn't recurse on the same stack.
     while (blockFuture.isDone() && !blockFuture.isCompletedExceptionally()) {
-      final boolean complete = handleLoadedBlock(requestState, blockFuture.join());
-      if (complete) {
+      if (blockFuture.join()) {
         return completedFuture(requestState);
       }
-      blockFuture = requestState.loadNextBlock();
+      blockFuture = processNextBlock(requestState);
     }
     return blockFuture.thenCompose(
-        maybeBlock ->
-            handleLoadedBlock(requestState, maybeBlock)
-                ? completedFuture(requestState)
-                : sendNextBlock(requestState));
+        complete -> complete ? completedFuture(requestState) : sendNextBlock(requestState));
+  }
+
+  private SafeFuture<Boolean> processNextBlock(final RequestState requestState) {
+    // Ensure blocks are loaded off of the event thread
+    return requestState
+        .loadNextBlock()
+        .thenCompose(block -> handleLoadedBlock(requestState, block));
   }
 
   /** Sends the block and returns true if the request is now complete. */
-  private boolean handleLoadedBlock(
+  private SafeFuture<Boolean> handleLoadedBlock(
       final RequestState requestState, final Optional<SignedBeaconBlock> block) {
-    block.ifPresent(requestState::sendBlock);
-    if (requestState.isComplete()) {
-      return true;
-    } else {
-      requestState.incrementCurrentSlot();
-      return false;
-    }
+    return block
+        .map(requestState::sendBlock)
+        .orElse(SafeFuture.COMPLETE)
+        .thenApply(
+            __ -> {
+              if (requestState.isComplete()) {
+                return true;
+              } else {
+                requestState.incrementCurrentSlot();
+                return false;
+              }
+            });
   }
 
   private class RequestState {
```

Comment:
Looks like `RequestState.currentSlot` and `RequestState.remainingBlocks` should be `volatile`.

Reply:
They don't need to be volatile because there's only one thread at a time accessing them (ie we load one block at a time).  Even though those actions might wind up happening on different threads, `CompletableFuture` guarantees that anything that happens in a thread before completing a future `happens-before` the completion stage runs (which applies even if it's run with one of the Async methods).  So even as we jump threads, we're guaranteed to see the most up to date values of those variables.